getStartsAndEnds.js

```diff
@@ -0,0 +1,31 @@
+/**
+ * Created by dusting on 11/4/15.
+ */
+var _ = require('underscore');
+
+var internal = exports.internal = {};
```

Comment:
What is the algorithmic complexity of this "Find start and end nodes" part of your algorithm?